Sample scripts to automate scaleable runners

bryanmacfarlane opened the pull request:
👉 [overview here](https://github.com/actions/runner/blob/automate/docs/automate.md) 👈

joshmgross commented on docs/automate.md:
```diff
@@ -0,0 +1,26 @@
+
+### Step 1: export PAT
+
+Create a GitHub PAT and export it before running the script
```
We might want to clarify the scopes required for the PAT so that users don't have to look at the API docs to know. I also don't think we need to specify a "GitHub PAT".

```suggestion
Create a PAT with `repo` scopes for repo runners and `admin:org` scopes for org runners. Export this PAT as an environment variable before running the script.
```

bryanmacfarlane replied:
Cool.  Is there somewhere in docs that clarifies the minimum scopes?  Ideally, I would just link to docs.  But agreed, good addition.

joshmgross commented on docs/automate.md:
````diff
@@ -0,0 +1,26 @@
+
+Repo level one liner; replace yourorg/yourrepo
+```bash
+curl https://raw.githubusercontent.com/actions/runner/automate/scripts/latest-svc.sh | bash -s yourorg/yourrepo
````
Users might be reluctant to pipe a script straight to bash, should we include instructions for downloading it and then running it as well?

bryanmacfarlane replied:
Done.  Added a link to sample script

joshmgross commented on scripts/latest-svc.sh:
```diff
@@ -0,0 +1,130 @@
+#
+# Examples:
+# RUNNER_CFG_PAT=<yourPAT> ./nix-svc.sh myuser/myrepo
+# RUNNER_CFG_PAT=<yourPAT> ./nix-svc.sh myorg
```
```suggestion
# RUNNER_CFG_PAT=<yourPAT> ./latest-svc.sh myuser/myrepo
# RUNNER_CFG_PAT=<yourPAT> ./latest-svc.sh myorg
```

joshmgross replied:
This was fixed as I commented

joshmgross commented on scripts/create-latest-svc.sh:
```diff
@@ -0,0 +1,130 @@
+#
+
+runner_scope=${1}
+runner_name=${2:-$(hostname)}
```
If I'm spinning up a lot of VMs to run this script and create runners, are we guaranteed to have a unique hostname for each one?

bryanmacfarlane replied:
That's why we take it as arg2.  But the vast majority of the time, hostname being the default is correct.

joshmgross replied:
I missed that, was looking at the Examples in this script file and didn't look at usage.

Should we have flags for those inputs? I.e. what if I want to specify User but not Name

Also thinking ahead to labels inputs, relying on the order of the arguments might not hold up

bryanmacfarlane replied:
It's simple and customary for bash scripting.  I thought about the user issue - it's a real edge and in that case you put $(hostname) in front of it.  If we did flags I think the sample script would get bogged down with a bunch of parsing logic and folks won't be able to see the forest through the weeds of the script.  That is, unless someone knows of a trivial (works everywhere) cmd line parsing bash technique ...

joshmgross commented on scripts/create-latest-svc.sh:
```diff
@@ -0,0 +1,130 @@
+tar xzf "./${runner_file}" -C runner
+
+# export of pass
+sudo chown -R $svc_user ./runner
```
Does this assume passwordless sudo? Should we have an option for a user to pass in a password?

bryanmacfarlane replied:
That's on the user.   Most init script and VMs should run with passwordless sudo, if not, they will get prompted per design.  I think elevating needs to be explicit.

joshmgross commented on scripts/create-latest-svc.sh:
```diff
@@ -0,0 +1,130 @@
+runner_url="https://github.com/${runner_scope}"
+echo
+echo "Configuring ${runner_name} @ $runner_url"
+echo "./config.sh --unattended --url $runner_url --token $RUNNER_TOKEN --name $runner_name"
```
`$RUNNER_TOKEN` is a secret value (works for up to an hour). Do we need to echo it out?

bryanmacfarlane replied:
I think in the script case it runs in seconds and docs cover that so I'm fine not.

TingluoHuang replied:
RUNNER_TOKEN is the pool admin token, it has permission to remove other runners.
I don't think we should print it in any log.

bryanmacfarlane replied:
fixing.  accidentally left in when I was debugging a problem.

TingluoHuang commented on docs/automate.md:
```diff
@@ -0,0 +1,57 @@
+
+:point_right: [Sample script here](../scripts/remove-svc.sh) :point_left:
+
+Repo level one liner.  NOTE: replace with yourorg/yourrepo (repo level) or just yourorg (org level) 
```
`Run as a one-line` instead of `Repo level one liner.` to match config ☝️ ☝️ ?

bryanmacfarlane replied:
I fixed them to all be consistent.  `one-liner`

TingluoHuang commented on docs/automate.md:
````diff
@@ -0,0 +1,57 @@
+curl -s https://raw.githubusercontent.com/actions/runner/automate/scripts/remove-svc.sh | bash -s yourorg/yourrepo
+```
+
+### Delete an offline runner
````
do we need this? we are going to clean up dormant runners if the runner offline for 30 days.

joshmgross replied:
Users might not want that offline runner showing up in the settings UI for 30 days

TingluoHuang replied:
my question is more like when should we expect the customer to use this?
I can imagine in an autoscale VM/Kubernetes scenario, the customer will use the config script to set up a runner when creating VM/container and use the un-config script to un-config runner before we delete the VM/container.
If the VM/container crash or shutdown unexpected, we will use the delete offline script for cleanup. 

@bryanmacfarlane is my understand right?

bryanmacfarlane replied:
It's used by the same people that need the delete api (as opposed to the cli to remove).  This is used for outside/in automation where the runner goes away.  The customers I've interacted with so far for org level runners were using it for container / K8s runners where there's just an entry point.  The container is stopped outside in and then they to delete it after it goes away.

TingluoHuang commented on scripts/create-latest-svc.sh:
```diff
@@ -0,0 +1,130 @@
+echo "Generating a registration token..."
+
+# if the scope has a slash, it's an repo runner
+base_api_url="https://api.github.com/orgs"
```
do we want to support GHES?

bryanmacfarlane replied:
Maybe later.  These are also sample scripts so they can copy and alter per instructions.  I want to walk a fine line between the samples being useful / illustrative but not too complex where it addresses all scenarios.

TingluoHuang commented on scripts/create-latest-svc.sh:
```diff
@@ -0,0 +1,130 @@
+
+latest_version_label=$(curl -s -X GET 'https://api.github.com/repos/actions/runner/releases/latest' | jq -r '.tag_name')
+latest_version=$(echo ${latest_version_label:1})
+runner_file="actions-runner-${runner_plat}-x64-${latest_version}.tar.gz"
```
what about ARM and ARM64? 😄

bryanmacfarlane replied:
Later.  We also need to do windows.  This first set of sample scripts illustrate linux and osx.  The sample should help others getting started as well.  See my point above

TingluoHuang commented on scripts/create-latest-svc.sh:
```diff
@@ -0,0 +1,130 @@
+svc_user=${3:-$USER}
+
+echo "Configuring runner @ ${runner_scope}"
+sudo echo
```
why do we need to `sudo echo`, to avoid prompt downstream?

bryanmacfarlane replied:
Yeah, I was trying to prompt early so they don't kick it off, look away only to find a prompt later.

TingluoHuang commented on scripts/create-latest-svc.sh:
```diff
@@ -0,0 +1,130 @@
+echo
+echo "Configuring ${runner_name} @ $runner_url"
+echo "./config.sh --unattended --url $runner_url --token $RUNNER_TOKEN --name $runner_name"
+sudo -E -u ${svc_user} ./config.sh --unattended --url $runner_url --token $RUNNER_TOKEN --name $runner_name
```
do we need `--replace`?

bryanmacfarlane replied:
I decided not too as it's dangerous.  Seems like an interactive choice.  If the script is explicit and it already exists, then they can fix it before running again or fix their outside / in automation.

TingluoHuang commented on scripts/create-latest-svc.sh:
```diff
@@ -0,0 +1,130 @@
+#---------------------------------------
+echo
+echo "Configuring as a service ..."
+./svc.sh install ${svc_user}
```
do we need to `sudo` for Linux?

bryanmacfarlane replied:
Yeah.  Fixed.

TingluoHuang commented on scripts/remove-svc.sh:
```diff
@@ -0,0 +1,67 @@
+echo
+echo "Uninstall the service ..."
+pushd ./runner
+./svc.sh stop
```
do we need `sudo`?

bryanmacfarlane replied:
Yeah,  Fixed.  only on Linux.  (osx does a user level launchd)

bryanmacfarlane commented:
@TingluoHuang - Josh looked and approved.  Can you have a look?  Or suggest someone else?  I want to make sure these are ready to go with org level runner release.

bryanmacfarlane commented:
I verified on linux (Debian flavors) and osx with repo and org level ...

TingluoHuang commented on scripts/create-latest-svc.sh:
```diff
@@ -0,0 +1,135 @@
+    base_api_url="https://api.github.com/repos"
+fi
+
+export RUNNER_TOKEN=$(curl -s -X POST ${base_api_url}/${runner_scope}/actions/runners/registration-token -H "accept: application/vnd.github.everest-preview+json" -H "authorization: token ${RUNNER_CFG_PAT}" | jq -r '.token')
```
in the error case, like my PAT doesn't have enough permission, will the customer have enough info to debug?

bryanmacfarlane replied:
It returns what the API returns since it's a simple script.  I think it's fine unless you have something more elegant in mind.  I don't think we can tell why it's access denied - it just is.

TingluoHuang commented on scripts/delete.sh:
```diff
@@ -0,0 +1,83 @@
+#     ./delete.sh scope name
+#
+#      scope required  repo (:owner/:repo) or org (:organization)
+#      name  optional  defaults to hostname.  name to delete
```
name is required too, i think.

bryanmacfarlane replied:
Fixed.

TingluoHuang commented on scripts/delete.sh:
```diff
@@ -0,0 +1,83 @@
+# The caller should have already ensured the runner is gone and/or stopped
+#
+# Examples:
+# RUNNER_CFG_PAT=<yourPAT> ./delete.sh myuser/myrepo
```
the example doesn't contain runner_name input

bryanmacfarlane replied:
fixed

TingluoHuang commented on scripts/delete.sh:
```diff
@@ -0,0 +1,83 @@
+#--------------------------------------
+# Ensure offline
+#--------------------------------------
+runner_status=$(curl -s -X GET ${base_api_url}/${runner_scope}/actions/runners  -H "accept: application/vnd.github.everest-preview+json" -H "authorization: token ${RUNNER_CFG_PAT}" \
```
what about paging?
can we query `actions/runners/runner1`?

bryanmacfarlane replied:
made the simple sample script work with 100.  Added note and TODO that more complex cases can loop on pages.

TingluoHuang commented on scripts/delete.sh:
```diff
@@ -0,0 +1,83 @@
+        | jq -M -j ".runners | .[] | [select(.name == \"${runner_name}\")] | .[0].status")
+
+if [ -z "${runner_status}" ]; then 
+    fatal "Could not find runner with name ${runner_name}"
```
do we have to fetal? should we just return since the runner has gone?

bryanmacfarlane replied:
I added fatal.  It prints an error and returns exiting the script with 1

TingluoHuang commented on scripts/delete.sh:
```diff
@@ -0,0 +1,83 @@
+#--------------------------------------
+# Get id of runner to remove
+#--------------------------------------
+runner_id=$(curl -s -X GET ${base_api_url}/${runner_scope}/actions/runners  -H "accept: application/vnd.github.everest-preview+json" -H "authorization: token ${RUNNER_CFG_PAT}" \
```
any way to get runner_id and runner_status in one curl request?

bryanmacfarlane replied:
I could potentially refactor to save off contents of get and then apply different jq filters independently however I think this is more readable as a simple script.

TingluoHuang commented on scripts/delete.sh:
```diff
@@ -0,0 +1,83 @@
+#--------------------------------------
+# Remove the runner
+#--------------------------------------
+curl -s -X DELETE ${base_api_url}/${runner_scope}/actions/runners/${runner_id} -H "authorization: token ${RUNNER_CFG_PAT}"
```
do we need any preview HTTP header like `application/vnd.github.everest-preview+json`?

bryanmacfarlane replied:
I think we're supposed to doing preview periods.  https://developer.github.com/changes/2/#--creating-a-repository-dispatch-event-now-official-part-of-github-api

TingluoHuang replied:
the reason I asked is that you have the preview header in the get runner endpoint and get registration token endpoint call.

TingluoHuang commented on scripts/remove-svc.sh:
```diff
@@ -0,0 +1,76 @@
+fi 
+${prefix}./svc.sh stop
+${prefix}./svc.sh uninstall
+./config.sh remove --token $REMOVE_TOKEN
```
should we `sudo ./config.sh remove` like what we did during config?
if the runner folder belongs to a different svc user, config.sh may fail without sudo because the current user doesn't have permission.

bryanmacfarlane replied:
fixed

bryanmacfarlane commented:
@TingluoHuang I addressed most feedback and left some comments and TODO for further work later.  I think the scripts as is help get folks going (I know it would have helped with a few customers I interacted with directly this week).
